music21/metadata/__init__.py

```diff
@@ -232,6 +415,533 @@ def __init__(self, *args, **keywords):
             if attr in keywords:
                 setattr(self, attr, keywords[attr])
 
+
+# -----------------------------------------------------------------------------
+# PUBLIC APIs:
+#   {get,add,set}{Item,Items}
+#   {get,add,set}Personal{Item,Items}
+#   getAllItems(skipContributors=False)
+#   getAllContributorItems
+
+
+    # never returns a list, always returns None, or the first item
+    def getItem(self,
+                key: str,
+                namespace: Optional[str] = None) -> Optional[Any]:
```

Comment:
I believe I have patterned my new API after the existing API pretty
effectively (particularly after the existing contributor APIs, which have
multiple values like all metadata does in my new API), with a few additions
to cover some new cases.

Here is my comparison of the existing API vs. my new API, with notes (search
for "Notes:") about how I could make things feel even more like the existing
API if you would like.

Existing APIs:

addContributor(c: Contributor) adds a contributor to the role list for c.role.
Left in place for backward compatibility; only works with the existing short
list of contributor roles.
Replaced by:
addItem(name: str, item: Any, namespace: Optional[str]=None) which works for
any metadata item (not just for contributors).

getContributorsByRole(role: str) gets a list of contributor for the role.
Left in place for backward compatibility; only works with the existing short
list of contributor roles.
Replaced by:
getItems(name: str, namespace: Optional[str]=None) which returns a list, and
works for any metadata item (not just for contributors).

There is no existing setContributorsByRole, but there is a "replacement":
setItems(name: str, item: Any, namespace: Optional[str]=None) which works for
any metadata item (not just for contributors).

setWorkId(idStr: str, value: Union[str, Text])
Left in place for backward compatibility; only works with the existing short
list of workIds.
Replaced by:
setItem(name: str, item: Any)

There is no existing getWorkId, but there is a "replacement":
getItem(name: str, namespace: Optional[str]=None)

Existing properties:

alternativeTitle et al: sets/gets a particular workId
ota et al: sets/gets a particular workId by abbreviation
I currently have left these in place and have not extended them beyond the
existing short list of workIds and abbreviations.
**Notes:** I would be happy to add more property setters/getters to include all the
new uniqueNames, but not the new abbreviations (since they are not unique).  If
I do so, I would want to return the full value type (e.g. Text, not str) from
all of them, including the existing workIds, which would break backward
compatibility.

composer et al (contributor role property setters and getters) sets/gets the
first contributor of a particular role.  I currently have left these in place
and have not extended them beyond the existing short list of contributor roles.
**Notes:** I would be happy to add more property setters/getters to include all the
new contributor role uniqueNames. If I do so, I would want to return Contributor
rather than str from all of them, including the existing short list of contributor
roles, which would break backward compatibility.

composers et al (plural contributor role property setters and getters).  I
currently have left these in place and have not extended them beyond the
existing short list of contributor roles.
**Notes:** I would be happy to add more plural property setters and getters to
include all the new contributor role uniqueNames.  If I do so, I would want to
return Contributor rather than str from all of them, including the existing
short list of contributor roles, which would break backward compatibility.
**Notes:** If I add plural properties for all the new contributor roles, I should
also add plural properties for all the new metadata uniqueNames as well (e.g.
md.abstracts).

New APIs for which there is no existing equivalent:

There is no existing way to get/set multiple values for a non-contributor, or
to add to a list of values for a non-contributor.
This is a new feature, supported by these new APIs:
getItems, setItems, and addItem
These APIs work for contributors as well, with getItems and addItem replacing
getContributorsByRole and addContributor, and with setItems taking the place
of the non-existent setContributorsByRole, as I mentioned above.

There is no existing way to get/set any metadata that isn't in the officially
supported list of names.
This is a new feature supported by these new APIs:
getPersonalItem/setPersonalItem
getPersonalItems/setPersonalItems
**Notes:** If you have a better naming idea than "Personal" for "randomly named
metadata item that we've never heard of before, and is probably specific to
one particular score-annotating client or group of clients", I'm all ears.

APIs I haven't figured out yet:
search(): TBD (currently backward compatible, only supports existing short list)
RichMetadata: TBD (currently backward compatible, only supports existing short list)

Reply:
Another thought after some poking around JavaScript docs (I'm not a JS guy): I intended the word "Item" to mean a singular piece of metadata, and I apparently ran into some names that are JS-like.  Perhaps a rename?  How about:

md.getItem() => md.get()
md.setItem() => md.set()
md.addItem() => md.add()
md.getItems() => md.getList()
md.setItems() => md.setList()

md.getPersonalItem() => md.getPersonal()
md.setPersonaltem() => md.setPersonal()
md.addPersonalItem() => md.addPersonal()
md.getPersonalItems() => md.getPersonalList()
md.setPersonalItems() => md.setPersonalList()

Reply:
I like these names.  Though I wonder if we want to say getAll() and setAll() instead of List, since I think we'll likely want to return them as a tuple or something like a MetadataList or something like that, since `md.getList()[2] = Text('Myke')` might be something we want to discourage.

Reply:
Sounds good!  Thanks.  I will also change "Personal" to "Custom" throughout.  It's a far better name.

Reply:
Done (and returning a tuple from getAll).